### PR TITLE
Update py-quantities

### PR DESCRIPTION
**Pull request by Moritz-Alexander-Kern**
- add version 0.14.1

### CODE REVIEW

File: var/spack/repos/builtin/packages/py-quantities/package.py [hunk 1]
```diff
@@ -11,20 +11,26 @@ class PyQuantities(PythonPackage):
 
     homepage = "https://python-quantities.readthedocs.org"
     pypi = "quantities/quantities-0.12.1.tar.gz"
-    maintainers("apdavison")
+    maintainers = ["apdavison"]
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/py-quantities/package.py]: Please restore the new directive (see https://spack.readthedocs.io/en/latest/packaging_guide.html#maintainers).

**Review by tldahlgren — DISMISSED** [review]
Confirmed the version sha256 but need to restore the maintainers directive (syntax).

Also deferring to the python package gurus for reviewing the rest of the PR.

File: var/spack/repos/builtin/packages/py-quantities/package.py [hunk 2]
```diff
@@ -11,20 +11,26 @@ class PyQuantities(PythonPackage):
...
     version("0.13.0", sha256="0fde20115410de21cefa786f3aeae69c1b51bb19ee492190324c1da705e61a81")
     version("0.12.5", sha256="67546963cb2a519b1a4aa43d132ef754360268e5d551b43dd1716903d99812f0")
     version("0.12.4", sha256="a33d636d1870c9e1127631185d89b0105a49f827d6aacd44ad9d8f151f331d8b")
     version("0.12.3", sha256="582f3c7aeba897846761e966615e01202a5e5d06add304492931b05085d19883")
 
     depends_on("python@2.7.0:2.7,3.4:3.7", type=("build", "run"), when="@0.12.3")
     depends_on("python@2.7.0:2.7,3.4:3.8", type=("build", "run"), when="@0.12.4:0.12.5")
-    depends_on("python@3.7:3.10", type=("build", "run"), when="@0.13:")
+    depends_on("python@3.7:3.10", type=("build", "run"), when="@0.13.0:0.13.1")
+    depends_on("python@3.8:3.11", type=("build", "run"), when="@0.14:")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-quantities/package.py]: Where do the upper bounds come from?

File: var/spack/repos/builtin/packages/py-quantities/package.py [hunk 3]
```diff
@@ -11,20 +11,26 @@ class PyQuantities(PythonPackage):
...
     # pip silently replaces distutils with setuptools
+    # py-setuptools@:61 doesn't support PEP 621
+    depends_on("py-setuptools@61:", type="build", when="@0.14:")
     depends_on("py-setuptools", type="build")
+    depends_on("py-setuptools-scm", type="build")
 
     depends_on("py-numpy@1.8.2:1.16", type=("build", "run"), when="@0.12.3")
     depends_on("py-numpy@1.8.2:1.17", type=("build", "run"), when="@0.12.4:0.12")
-    depends_on("py-numpy@1.16:", type=("build", "run"), when="@0.13.0:")
+    depends_on("py-numpy@1.16:", type=("build", "run"), when="@0.13.0:0.13.1")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-quantities/package.py]:
```suggestion
    depends_on("py-numpy@1.16:", type=("build", "run"), when="@0.13")
```

File: var/spack/repos/builtin/packages/py-quantities/package.py [hunk 4]
```diff
@@ -11,20 +11,26 @@ class PyQuantities(PythonPackage):
...
     depends_on("python@2.7.0:2.7,3.4:3.8", type=("build", "run"), when="@0.12.4:0.12.5")
-    depends_on("python@3.7:3.10", type=("build", "run"), when="@0.13:")
+    depends_on("python@3.7:3.10", type=("build", "run"), when="@0.13.0:0.13.1")
+    depends_on("python@3.8:3.11", type=("build", "run"), when="@0.14:")
 
     # pip silently replaces distutils with setuptools
+    # py-setuptools@:61 doesn't support PEP 621
+    depends_on("py-setuptools@61:", type="build", when="@0.14:")
     depends_on("py-setuptools", type="build")
+    depends_on("py-setuptools-scm", type="build")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-quantities/package.py]:
```suggestion
    depends_on("py-setuptools-scm+toml", type="build")
```

**Review by adamjstewart — COMMENTED** [review]
I prefer to sort dependencies from newest to oldest like we do with versions

File: var/spack/repos/builtin/packages/py-quantities/package.py (hunk 2 above)
Comment (apdavison) [reply var/spack/repos/builtin/packages/py-quantities/package.py]: The upper bounds generally reflect the versions of Python that we run/ran CI testing for. If this is too restrictive, they could be removed.

**Review by apdavison — APPROVED** [review]

File: var/spack/repos/builtin/packages/py-quantities/package.py (hunk 1 above)
Comment (Moritz-Alexander-Kern) [reply var/spack/repos/builtin/packages/py-quantities/package.py]: Thanks for catching this, reverted to:  `maintainers("apdavison")`

**Comment by Moritz-Alexander-Kern** [comment]
> I prefer to sort dependencies from newest to oldest like we do with versions

Sorted dependencies from newest to oldest.

File: var/spack/repos/builtin/packages/py-quantities/package.py (hunk 2 above)
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-quantities/package.py]: I usually don't add Python version ranges unless:

1. the library is tightly coupled to the CPython API and has separate wheels for each Python version on PyPI, or
2. you know for a fact that newer versions of Python won't work.

We frequently deprecate and remove EOL Python versions so the older versions of this library would have to be removed too.

Comment (Moritz-Alexander-Kern) [reply var/spack/repos/builtin/packages/py-quantities/package.py]: Removed upper bounds for python.

File: var/spack/repos/builtin/packages/py-quantities/package.py [hunk 5]
```diff
@@ -13,18 +13,23 @@ class PyQuantities(PythonPackage):
...
-    depends_on("python@3.7:3.10", type=("build", "run"), when="@0.13:")
+    depends_on("python@3.8:", type=("build", "run"), when="@0.14:")
+    depends_on("python@3.7:", type=("build", "run"), when="@0.13.0:0.13.1")
+    depends_on("python@2.7.0:", type=("build", "run"), when="@0.12.3:0.12.5")
 
     # pip silently replaces distutils with setuptools
+    # py-setuptools@:61 doesn't support PEP 621
+    depends_on("py-setuptools@61:", type="build", when="@0.14:")
     depends_on("py-setuptools", type="build")
+    depends_on("py-setuptools-scm+toml", type="build")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-quantities/package.py]:
```suggestion
    depends_on("py-setuptools-scm+toml", type="build", when="@0.14:")
```

File: var/spack/repos/builtin/packages/py-quantities/package.py [hunk 6]
```diff
@@ -13,18 +13,23 @@ class PyQuantities(PythonPackage):
...
     version("0.12.5", sha256="67546963cb2a519b1a4aa43d132ef754360268e5d551b43dd1716903d99812f0")
     version("0.12.4", sha256="a33d636d1870c9e1127631185d89b0105a49f827d6aacd44ad9d8f151f331d8b")
     version("0.12.3", sha256="582f3c7aeba897846761e966615e01202a5e5d06add304492931b05085d19883")
 
-    depends_on("python@2.7.0:2.7,3.4:3.7", type=("build", "run"), when="@0.12.3")
-    depends_on("python@2.7.0:2.7,3.4:3.8", type=("build", "run"), when="@0.12.4:0.12.5")
-    depends_on("python@3.7:3.10", type=("build", "run"), when="@0.13:")
+    depends_on("python@3.8:", type=("build", "run"), when="@0.14:")
+    depends_on("python@3.7:", type=("build", "run"), when="@0.13.0:0.13.1")
+    depends_on("python@2.7.0:", type=("build", "run"), when="@0.12.3:0.12.5")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-quantities/package.py]: Could remove these lines if you want, Spack only supports Python 3.7+ anyway

File: var/spack/repos/builtin/packages/py-quantities/package.py [hunk 7]
```diff
@@ -13,18 +13,23 @@ class PyQuantities(PythonPackage):
...
 
     # pip silently replaces distutils with setuptools
+    # py-setuptools@:61 doesn't support PEP 621
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-quantities/package.py]: Could remove these comments, the dependency is now explicit

File: var/spack/repos/builtin/packages/py-quantities/package.py (hunk 6 above)
Comment (Moritz-Alexander-Kern) [reply var/spack/repos/builtin/packages/py-quantities/package.py]: Thank you, I agree this enhances the clarity.

File: var/spack/repos/builtin/packages/py-quantities/package.py (hunk 7 above)
Comment (Moritz-Alexander-Kern) [reply var/spack/repos/builtin/packages/py-quantities/package.py]: Comments removed. :+1:

File: var/spack/repos/builtin/packages/py-quantities/package.py [hunk 8]
```diff
@@ -13,18 +13,19 @@ class PyQuantities(PythonPackage):
...
+    depends_on("py-setuptools@61:", type="build", when="@0.14:")
     depends_on("py-setuptools", type="build")
+    depends_on("py-setuptools-scm+toml", type="build", when="@0.14:")
 
-    depends_on("py-numpy@1.8.2:1.16", type=("build", "run"), when="@0.12.3")
+    depends_on("py-numpy@1.19:", type=("build", "run"), when="@0.14:")
+    depends_on("py-numpy@1.16:", type=("build", "run"), when="@0.13")
     depends_on("py-numpy@1.8.2:1.17", type=("build", "run"), when="@0.12.4:0.12")
-    depends_on("py-numpy@1.16:", type=("build", "run"), when="@0.13.0:")
+    depends_on("py-numpy@1.8.2:1.16", type=("build", "run"), when="@0.12.3")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-quantities/package.py:31]: Should these numpy upper bounds also be removed for the same reason the python upper bounds were removed?

Comment (apdavison) [reply var/spack/repos/builtin/packages/py-quantities/package.py:31]: I think this is a different case; recent NumPy releases have deprecated then removed a number of behaviours, and if I remember correctly these earlier versions of `quantities` used these behaviours, and so there would be errors if they were used with later versions of `numpy`.

**Review by adamjstewart — APPROVED** [review]

**Pull request merged**